CL-959 Add support for representativeness to domicile custom field

adessy opened the pull request:
Sorrry, didn't have time. Maybe later 😴 

(Watch out: seems tests are not running... probably bc I force-pushed.)

adessy committed:
CL-959 Add 'custom_field_option_id column' to 'areas' table

This also includes a data migration that creates custom field options for the domicile field based on the configured areas.

adessy committed:
CL-959 Add custom field logic to `Area`

Add the custom_field_option association to the area model and after_update/create callbacks to keep areas and their associated option in sync.

adessy committed:
CL-959 Add area logic to `CustomFieldOption`

Add the area association to the CustomFieldOption model and an after_update callbacks to update the area when the option is modified.

adessy committed:
CL-959 Create options when domicile field is created

adessy committed:
CL-959 Fix failing tests

`create(:custom_field, key: 'domicile', code: 'domicile')` creates an inconsistent domicile field where the input type is 'text' instead of 'select'.

adessy committed:
CL-959 Use code to check if custom field is about domicile

adessy committed:
CL-959 Convert area ids to option keys in counts_by_field_option

adessy committed:
CL-959 Move sync logic between areas and field options to the core

Move sync logic between areas and field options to the core for more consistency.

adessy committed:
CL-959 Re-implement the data migration as a background job

adessy committed:
CL-959 Refactoring area_spec.rb

adessy committed:
CL-959 Add tests to area_spec.rb

adessy committed:
CL-959 Add tests to custom_field_spec.rb

adessy committed:
CL-959 Add tests to custom_field_option_spec.rb

adessy committed:
CL-959 Update users' custom field values when an area-option is deleted

jamesspeake reviewed (commented):
I am happy with most of the changes (noting there is some tech debt to sort out later) - however, I'm not confident in the Async nature of the DB migration given I could not get it to work locally. @kogre is going to take a look and make a call between this one and the previous one that I have enabled in branch CL-1157

kogre commented:
I'm not sure we should go with this. The code is fine, and I'm sure it works, but I think that we want to keep our data layer as clean and normalized as possible. I don't think syncing options and areas through our application layer is a good idea, it will bite us sooner rather than later. Because data manipulations will happen behind the back of our application layer, because the syncing logic will fail or records will be created that skip the AR hooks, because the application will fail with an unrelated error and not perform the syncing.

Given the effort put into this, it must be that I don't understand well enough what pain this is creating today. We have a few places where we need to deal with this exceptional behavior (using areas instead of custom field options for the field), and while it's always nicer to not have these, I don't consider this a great pain or big complexity in the code today. It's all quite contained and small in terms of code. So it must be that for representation, this is a big pain, and I'm not getting yet why.

I understand this is blocking the release of representation, so syncing today to understand how urgent it is to get this out.

adessy committed:
CL-959 Raise error when custom field is not domicile

I simplified to call `convert_area_ids_to_option_keys!` when `by_area_id` is true, even if the custom field is not domicile. In this case, `convert_area_ids_to_option_keys!` will raise an exception and make visible that we are not invoking the method correctly.

adessy committed:
CL-959 Improve tests

Remove some hardcoded values from tests.

adessy committed:
CL-959 Fix typo

adessy committed:
CL-959 Improve comments

adessy committed:
CL-959 Simplify `.counts_by_field_option` interface

Replace +by_area_id+ and +by_option_id+ boolean parameters by a single +by+ parameter that take a symbol. This removes the risk of having conflicting 'by' parameters.

adessy committed:
CL-959 Test indexing counts by area id

adessy committed:
Merge branch 'master' into CL-959/domicile-representativeness

# Conflicts:
#	back/app/models/custom_field.rb
#	back/db/schema.rb
#	back/spec/services/json_forms_user_override_spec.rb

adessy committed:
CL-959 Fix failing tests

adessy committed:
CL-959 Skip flaky test

adessy committed:
CL-959 Fix rubocop offenses

adessy committed:
CL-959 Fix rubocop offenses

adessy committed:
CL-959 Fix ordering of options for domicile field

It seems that `find_each` does not respect the default scope. As a result, areas were not always returned in the same order which messed up the ordering of options.

adessy committed:
Merge branch 'master' into CL-959/domicile-representativeness
